runtime/shared/src/main/scala/org/http4s/googleapis/runtime/auth/IdentityPoolCredentials.scala

```diff
@@ -0,0 +1,270 @@
+    * @param externalAccount
+    *   configuration for external account credentials generation
+    * @param scopes
+    *   scopes for authorization
+    *
+    * @return
+    *   a sts token for authorization skipping impersonation flow
+    */
+  private def withoutImpersonation[F[_]](
+      client: Client[F],
+      pid: String,
+      retrieveSubjectToken: F[String],
+      externalAccount: CredentialsFile.ExternalAccount,
+      scopes: Seq[String],
+  )(implicit F: Temporal[F]): F[GoogleCredentials[F]] = for {
+    // TODO: implement cache logic
+    _ <- F.ref(Option.empty[AccessToken])
+  } yield new GoogleCredentials[F] {
+    def projectId: String = pid
+    def get: F[AccessToken] = for {
+      sbjTkn <- retrieveSubjectToken
+      tkn <- GoogleOAuth2TokenExchange[F](client)
+        .stsToken(sbjTkn, externalAccount, scopes)
+      // If impersonation url is not available, end the flow and just use the STS access token for authorization.
+    } yield tkn
+  }
+
+  private def withImpersonation[F[_]: Files](
+      client: Client[F],
+      id: String,
+      impersonationURL: Uri,
+      fileSource: ExternalCredentialSource.File,
+      externalAccount: CredentialsFile.ExternalAccount,
+      scopes: Seq[String],
+  )(implicit F: Temporal[F]): F[GoogleCredentials[F]] = fileSource match {
+    case ExternalCredentialSource.File(file, format) =>
+      F.ifM(Files[F].exists(Path(file)))(
+        F.unit,
+        F.raiseError(new IOException(s"$file not found")),
+      ).flatMap(_ =>
+        withImpersonation(
+          client,
+          id,
+          impersonationURL,
+          subjectTokenFromFile(file, format),
+          externalAccount,
+          scopes,
+        ),
+      )
+  }
+  private def withImpersonation[F[_]](
+      client: Client[F],
+      id: String,
+      impersonationURL: Uri,
+      urlSource: ExternalCredentialSource.Url,
+      externalAccount: CredentialsFile.ExternalAccount,
+      scopes: Seq[String],
+  )(implicit F: Temporal[F]): F[GoogleCredentials[F]] = urlSource match {
+    case ExternalCredentialSource.Url(tknURL, headers, format) =>
+      F.fromEither(Uri.fromString(tknURL))
+        .flatMap(tknURL =>
+          withImpersonation(
+            client,
+            id,
+            impersonationURL,
+            subjectTokenFromUrl(client, tknURL, headers, format),
+            externalAccount,
+            scopes,
+          ),
+        )
+  }
+
+  /** The shared routine to fetch external credentials with service account impersonation
+    *
+    * @param id
+    *   GCP project id
+    * @param retrieveSubjectToken
+    *   abstract platform specific actions to retrieve subject token for token exchange
+    * @param externalAccount
+    *   configuration for external account credentials generation
+    * @param scopes
+    *   scopes for authorization
+    * @return
+    *   authorization token with impersonation
+    */
+  private def withImpersonation[F[_]](
```

Comment:
This is the shared flow to get GoogleCredentials __with__ service account impersonation.